lib/routes/account.js

```diff
@@ -406,52 +412,55 @@ module.exports = function (
+            maybeBlocked = e
+            return
+          }
+          throw e
+        }
+
         function readEmailRecord () {
           return db.emailRecord(email)
             .then(
               function (result) {
                 emailRecord = result
 
-                // Session token verification is only enabled for certain users during phased rollout.
-                // Even when it is enabled, we only do the email challenge if:
-                //  * the request wants keys, since unverified sessions are fine to use for e.g. oauth login.
-                //  * the email is verified, since content-server triggers a resend of the verification
-                //    email on unverified accounts, which doubles as sign-in confirmation.
-                if (! requestHelper.shouldEnableTokenVerification(emailRecord, config, request)) {
-                  tokenVerificationId = undefined
-                  mustVerifySession = false
-                  doSigninConfirmation = false
-                } else {
-                  // The user doesn't *have* to verify their session if they're not requesting keys,
-                  // but we still create it with a non-null tokenVerificationId, so it will still
-                  // be considered unverified.  This prevents the session from being used for sync
-                  // unless the user explicitly requests us to resend the confirmation email, and completes it.
-                  mustVerifySession = requestHelper.wantsKeys(request)
-                  doSigninConfirmation = mustVerifySession && emailRecord.emailVerified
-                }
-
-                if(email !== emailRecord.email) {
-                  customs.flag(request.app.clientAddress, {
-                    email: email,
-                    errno: error.ERRNO.INCORRECT_PASSWORD
-                  })
-                  throw error.incorrectPassword(emailRecord.email, email)
-                }
-
-                return checkPassword(emailRecord, authPW, request.app.clientAddress)
-                  .then(
-                    function (match) {
-                      if (! match) {
-                        throw error.incorrectPassword(emailRecord.email, email)
+                allowSigninUnblock = requestHelper.shouldEnableSigninUnblock(emailRecord, config, request)
+                if (allowSigninUnblock && unblockCode) {
+                  return db.consumeUnblockCode(emailRecord.uid, unblockCode)
+                    .then(
+                      function (code) {
+                        if (Date.now() - code.createdAt > unblockCodeLifetime) {
+                          log.info({
+                            op: 'Account.login.unblockCode.expired',
+                            uid: emailRecord.uid.toString('hex')
+                          })
+                          throw error.invalidUnblockCode()
+                        }
+                        didSigninUnblock = true
```

Comment:
Please emit a flowEvent indicating that unblock code was successfully consumed.

If someone logs in through this flow, it would be awesome to see it as a sequence of flow events like:

```
flow.begin
flow.engage
account.login.submit
error.114
account.login.sendUnblockCode
account.login.submit
account.login.confirmedUnblockCode
account.login
device.created
account.keyfetch
account.signed
```

(With the caveat that I just made up most of those names, but you get the idea).